src/ImageSharp/Formats/Bmp/BmpEncoderCore.cs

```diff
@@ -340,5 +363,51 @@ private void Write8Bit<TPixel>(Stream stream, ImageFrame<TPixel> image)
                 }
             }
         }
+
+        /// <summary>
+        /// Writes an 8 Bit gray image with a color palette. The color palette has 256 entry's with 4 bytes for each entry.
+        /// </summary>
+        /// <typeparam name="TPixel">The type of the pixel.</typeparam>
+        /// <param name="stream">The <see cref="Stream"/> to write to.</param>
+        /// <param name="image"> The <see cref="ImageFrame{TPixel}"/> containing pixel data.</param>
+        /// <param name="colorPalette">A byte span of size 1024 for the color palette.</param>
+        private void Write8BitGray<TPixel>(Stream stream, ImageFrame<TPixel> image, Span<byte> colorPalette)
+            where TPixel : struct, IPixel<TPixel>
+        {
+            // Create a color palette with 256 different gray values.
+            for (int i = 0; i <= 255; i++)
+            {
+                int idx = i * 4;
+                byte grayValue = (byte)i;
+                colorPalette[idx] = grayValue;
+                colorPalette[idx + 1] = grayValue;
+                colorPalette[idx + 2] = grayValue;
+
+                // Padding byte, always 0.
+                colorPalette[idx + 3] = 0;
+            }
+
+            stream.Write(colorPalette);
+
+            using (IMemoryOwner<byte> rowSpanBuffer = this.memoryAllocator.AllocateManagedByteBuffer(image.Width, AllocationOptions.Clean))
+            {
+                Span<byte> outputPixelRow = rowSpanBuffer.GetSpan();
+                for (int y = image.Height - 1; y >= 0; y--)
+                {
+                    Span<TPixel> inputPixelRow = image.GetPixelRowSpan(y);
+                    PixelOperations<TPixel>.Instance.ToGray8Bytes(
```

Comment:
Can't we replace this with `inputPixelRow.AsBytes().CopyTo(outputPixelRow)` here? Seems more explanatory about our intention.

Reply:
I can not find AsBytes(), is that an extension method im missing?

Reply:
Apologies, it's `MemoryMarshal.AsBytes(spanOfThings)` 😄

Reply:
i have changed that, please check it again.

Reply:
Looks very good, thanks!